File: Sources/MCPTools/Backends/BlenderBackend.swift
import Foundation
import MCPServer

/// Blender Lab (TCP, null-terminated JSON) vs community (WebSocket JSON text).
/// Port open is not enough — both default to :9876.
enum BlenderProtocol {
    static let defaultPorts: ClosedRange<UInt16> = 9876...9896

    static func encodeLab(code: String, strictJSON: Bool = true) throws -> Data {
        let payload = JSONValue.object([
            "type": .string("execute"),
            "code": .string(code),
            "strict_json": .bool(strictJSON),
        ])
        var data = try JSONEncoder().encode(payload)
        data.append(0)
        return data
    }

    static func decodeLab(_ data: Data) throws -> JSONValue {
        let slice = data.split(separator: 0, maxSplits: 1, omittingEmptySubsequences: true).first.map(Data.init) ?? data
        return try JSONDecoder().decode(JSONValue.self, from: slice)
    }

    static let pingCode = "result={'agentcontroller': True, 'objects': len(__import__('bpy').data.objects)}"

    static func isLabSuccess(_ value: JSONValue) -> Bool {
        let status = value["status"]?.stringValue?.lowercased()
        return status == "ok" || status == "success" || value["result"] != nil
    }
}

actor BlenderBackend {
    static let shared = BlenderBackend()

    struct Endpoint: Sendable, Equatable {
        var host: String
        var port: UInt16
        var kind: CapabilityRecord.Backend
        var pid: Int?
        var detail: String
    }

    private var communityWS: URLSessionWebSocketTask?

    func handshake(ports: ClosedRange<UInt16> = BlenderProtocol.defaultPorts) async -> [Endpoint] {
        await withTaskGroup(of: Endpoint?.self) { group in
            for port in ports {
                group.addTask {
                    await self.probe(port: port)
                }
            }
            var found: [Endpoint] = []
            for await item in group {
                if let item { found.append(item) }
            }
            return found.sorted { $0.port < $1.port }
        }
    }

    func execute(endpoint: Endpoint, code: String) async throws -> JSONValue {
        switch endpoint.kind {
        case .blenderLab:
            let payload = try BlenderProtocol.encodeLab(code: code)
            let data = try SocketProbe.roundTrip(
                host: endpoint.host,
                port: endpoint.port,
                payload: payload,
                timeoutMs: 8_000,
                readUntilNull: true
            )
            return try BlenderProtocol.decodeLab(data)
        case .blenderWS:
            return try await communityExecute(endpoint: endpoint, code: code)
        default:
            throw ToolError.actionFailed("Not a Blender endpoint")
        }
    }

    func sceneSnapshot(endpoint: Endpoint) async throws -> [RoutedRef] {
        let code = """
        import bpy
        result = [{'name': o.name, 'type': o.type, 'location': list(o.location)} for o in bpy.data.objects]
        """
        let raw = try await execute(endpoint: endpoint, code: code)
        let list = raw["result"] ?? raw
        let items: [JSONValue]
        if let array = list.arrayValue {
            items = array
        } else if let array = list["result"]?.arrayValue {
            items = array
        } else {
            items = []
        }
        return items.compactMap { item in
            let name = item["name"]?.stringValue ?? item.stringValue
            guard let name, !name.isEmpty else { return nil }
            let kind = item["type"]?.stringValue ?? "OBJECT"
            return RoutedRef.blender(name: name, kind: kind)
        }
    }

    private func probe(port: UInt16) async -> Endpoint? {
        // Lab first: a tiny execute. Community sockets will fail to parse this.
        if let endpoint = probeLab(port: port) { return endpoint }
        if let endpoint = await probeCommunity(port: port) { return endpoint }
        return nil
    }

    private func probeLab(port: UInt16) -> Endpoint? {
        guard let payload = try? BlenderProtocol.encodeLab(code: BlenderProtocol.pingCode) else { return nil }
        guard let data = try? SocketProbe.roundTrip(
            host: "127.0.0.1",
            port: port,
            payload: payload,
            timeoutMs: 250,
            readUntilNull: true
        ), !data.isEmpty else { return nil }
        guard let decoded = try? BlenderProtocol.decodeLab(data), BlenderProtocol.isLabSuccess(decoded) else {
            return nil
        }
        return Endpoint(host: "127.0.0.1", port: port, kind: .blenderLab, pid: nil, detail: "Blender Lab MCP (null-terminated JSON)")
    }

    private func probeCommunity(port: UInt16) async -> Endpoint? {
        guard let url = URL(string: "ws://127.0.0.1:\(port)") else { return nil }
        let task = URLSession.shared.webSocketTask(with: url)
        task.resume()
        let ping = "{\"type\":\"get_scene_info\"}"
        do {
            try await task.send(.string(ping))
            let message = try await withTimeout(ms: 250) {
                try await task.receive()
            }
            task.cancel(with: .goingAway, reason: nil)
            switch message {
            case .string(let text):
                if text.contains("error") && text.lowercased().contains("unknown") { return nil }
                return Endpoint(host: "127.0.0.1", port: port, kind: .blenderWS, pid: nil, detail: "Blender community WebSocket")
            case .data(let data):
                if let text = String(data: data, encoding: .utf8), text.contains("{") {
                    return Endpoint(host: "127.0.0.1", port: port, kind: .blenderWS, pid: nil, detail: "Blender community WebSocket")
                }
            @unknown default:
                return nil
            }
        } catch {
            task.cancel(with: .goingAway, reason: nil)
            return nil
        }
        return nil
    }

    private func communityExecute(endpoint: Endpoint, code: String) async throws -> JSONValue {
        guard let url = URL(string: "ws://\(endpoint.host):\(endpoint.port)") else {
            throw ToolError.actionFailed("Invalid Blender WebSocket URL")
        }
        let task = URLSession.shared.webSocketTask(with: url)
        task.resume()
        defer { task.cancel(with: .goingAway, reason: nil) }
        let body = JSONValue.object([
            "type": .string("execute_code"),
            "params": .object(["code": .string(code)]),
        ])
        let data = try JSONEncoder().encode(body)
        guard let text = String(data: data, encoding: .utf8) else {
            throw ToolError.actionFailed("Failed to encode Blender payload")
        }
        try await task.send(.string(text))
        let message = try await withTimeout(ms: 8_000) { try await task.receive() }
        switch message {
        case .string(let s):
            return try JSONDecoder().decode(JSONValue.self, from: Data(s.utf8))
        case .data(let d):
            return try JSONDecoder().decode(JSONValue.self, from: d)
        @unknown default:
            throw ToolError.actionFailed("Unexpected Blender WebSocket frame")
        }
    }
}

func withTimeout<T: Sendable>(ms: Int, body: @escaping @Sendable () async throws -> T) async throws -> T {
    try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await body() }
        group.addTask {
            try await Task.sleep(nanoseconds: UInt64(ms) * 1_000_000)
            throw ToolError.timedOut("backend handshake")
        }
        guard let result = try await group.next() else {
            throw ToolError.timedOut("backend handshake")
        }
        group.cancelAll()
        return result
    }
}

File: Windows/AgentController.Windows/Backends/BlenderBackend.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json.Nodes;

namespace AgentController.Windows.Backends;

internal sealed record BlenderEndpoint(string Host, int Port, string Kind, string Detail);

internal static class BlenderBackend
{
    private const string PingCode = "result={'agentcontroller': True, 'objects': len(__import__('bpy').data.objects)}";

    internal static byte[] EncodeLab(string code)
    {
        var json = new JsonObject { ["type"] = "execute", ["code"] = code, ["strict_json"] = true }.ToJsonString();
        var bytes = Encoding.UTF8.GetBytes(json);
        var framed = new byte[bytes.Length + 1];
        Buffer.BlockCopy(bytes, 0, framed, 0, bytes.Length);
        return framed;
    }

    internal static JsonObject DecodeLab(byte[] data)
    {
        var end = Array.IndexOf(data, (byte)0);
        var slice = end >= 0 ? data.AsSpan(0, end) : data.AsSpan();
        return JsonNode.Parse(Encoding.UTF8.GetString(slice)) as JsonObject ?? [];
    }

    internal static bool IsLabSuccess(JsonObject value)
    {
        var status = value["status"]?.GetValue<string>()?.ToLowerInvariant();
        return status is "ok" or "success" || value["result"] is not null;
    }

    internal static List<BlenderEndpoint> Handshake()
    {
        var found = new List<BlenderEndpoint>();
        for (var port = 9876; port <= 9896; port++)
        {
            if (ProbeLab(port) is { } lab) { found.Add(lab); continue; }
            if (ProbeWs(port) is { } ws) found.Add(ws);
        }
        return found;
    }

    internal static JsonNode Execute(BlenderEndpoint endpoint, string code)
    {
        if (endpoint.Kind == "bpy-lab")
        {
            var data = RoundTrip(endpoint.Host, endpoint.Port, EncodeLab(code), 8000, untilNull: true);
            return DecodeLab(data);
        }
        using var ws = new ClientWebSocket();
        ws.ConnectAsync(new Uri($"ws://{endpoint.Host}:{endpoint.Port}"), CancellationToken.None).GetAwaiter().GetResult();
        var payload = Encoding.UTF8.GetBytes(new JsonObject
        {
            ["type"] = "execute_code",
            ["params"] = new JsonObject { ["code"] = code }
        }.ToJsonString());
        ws.SendAsync(payload, WebSocketMessageType.Text, true, CancellationToken.None).GetAwaiter().GetResult();
        var buffer = new byte[1 << 16];
        var result = ws.ReceiveAsync(buffer, CancellationToken.None).GetAwaiter().GetResult();
        return JsonNode.Parse(Encoding.UTF8.GetString(buffer, 0, result.Count)) ?? new JsonObject();
    }

    internal static List<JsonObject> SceneObjects(BlenderEndpoint endpoint)
    {
        var raw = Execute(endpoint, "import bpy\nresult=[{'name': o.name, 'type': o.type} for o in bpy.data.objects]");
        var result = raw["result"] ?? raw;
        if (result is JsonArray array)
            return array.OfType<JsonObject>().Where(item => item["name"] is not null).ToList();
        return [];
    }

    private static BlenderEndpoint? ProbeLab(int port)
    {
        try
        {
            var data = RoundTrip("127.0.0.1", port, EncodeLab(PingCode), 250, untilNull: true);
            var decoded = DecodeLab(data);
            return IsLabSuccess(decoded)
                ? new("127.0.0.1", port, "bpy-lab", "Blender Lab MCP (null-terminated JSON)")
                : null;
        }
        catch { return null; }
    }

    private static BlenderEndpoint? ProbeWs(int port)
    {
        try
        {
            using var ws = new ClientWebSocket();
            using var cts = new CancellationTokenSource(250);
            ws.ConnectAsync(new Uri($"ws://127.0.0.1:{port}"), cts.Token).GetAwaiter().GetResult();
            var payload = Encoding.UTF8.GetBytes("""{"type":"get_scene_info"}""");
            ws.SendAsync(payload, WebSocketMessageType.Text, true, cts.Token).GetAwaiter().GetResult();
            var buffer = new byte[4096];
            var result = ws.ReceiveAsync(buffer, cts.Token).GetAwaiter().GetResult();
            var text = Encoding.UTF8.GetString(buffer, 0, result.Count);
            if (!text.Contains('{')) return null;
            return new("127.0.0.1", port, "bpy-ws", "Blender community WebSocket");
        }
        catch { return null; }
    }

    private static byte[] RoundTrip(string host, int port, byte[] payload, int timeoutMs, bool untilNull)
    {
        using var client = new System.Net.Sockets.TcpClient();
        var connect = client.ConnectAsync(host, port);
        if (!connect.Wait(timeoutMs) || !client.Connected) throw new TimeoutException();
        using var stream = client.GetStream();
        stream.ReadTimeout = timeoutMs;
        stream.WriteTimeout = timeoutMs;
        stream.Write(payload);
        var buffer = new MemoryStream();
        var chunk = new byte[16384];
        while (true)
        {
            var n = stream.Read(chunk, 0, chunk.Length);
            if (n <= 0) break;
            buffer.Write(chunk, 0, n);
            if (untilNull && buffer.ToArray().Contains((byte)0)) break;
            if (!untilNull) break;
        }
        if (buffer.Length == 0) throw new TimeoutException();
        return buffer.ToArray();
    }
}
